### PR TITLE
feat(aws): Batch services

### PR DESCRIPTION
**Pull request by tjholm**
Implementation of Nitric batch services on AWS

### CODE REVIEW

File: cloud/aws/deploy/batch.go [hunk 1]
```diff
@@ -15,13 +15,348 @@
...
+	a.BatchSecurityGroup, err = awsec2.NewSecurityGroup(ctx, "batch-sg", &awsec2.SecurityGroupArgs{
+		VpcId: vpcId,
+		Egress: awsec2.SecurityGroupEgressArray{
+			awsec2.SecurityGroupEgressArgs{
+				Protocol: pulumi.String("-1"),
+				FromPort: pulumi.Int(0),
+				ToPort:   pulumi.Int(0),
+				// Still need public internet access for batch jobs by default
+				// TODO: Allow restriction via config
+				CidrBlocks: pulumi.StringArray{pulumi.String("0.0.0.0/0")},
```
Comment (HomelessDinosaur) [inline cloud/aws/deploy/batch.go:74]: Can this be resolved?

File: cloud/aws/deploy/batch.go [hunk 2]
```diff
@@ -15,13 +15,348 @@
...
+		AssumeRolePolicy: pulumi.String(`{
+			"Version": "2012-10-17",
+			"Statement": [
+				{
+					"Action": "sts:AssumeRole",
+					"Principal": {
+						"Service": "batch.amazonaws.com"
+					},
+					"Effect": "Allow",
+					"Sid": ""
```
Comment (HomelessDinosaur) [inline cloud/aws/deploy/batch.go:118]:
```suggestion
					"Sid": "BatchExecutionRole"
```

File: cloud/aws/deploy/batch.go [hunk 3]
```diff
@@ -15,13 +15,348 @@
...
+		AssumeRolePolicy: pulumi.String(`{
+			"Version": "2012-10-17",
+			"Statement": [
+				{
+					"Action": "sts:AssumeRole",
+					"Principal": {
+						"Service": "ecs-tasks.amazonaws.com"
+					},
+					"Effect": "Allow",
+					"Sid": ""
```
Comment (HomelessDinosaur) [inline cloud/aws/deploy/batch.go:220]:
```suggestion
					"Sid": "BatchJobRole"
```

File: cloud/aws/deploy/batch.go [hunk 4]
```diff
@@ -15,13 +15,348 @@
...
+		"dynamodb:ListTables",
+		"s3:ListAllMyBuckets",
+		"tag:GetResources",
+		"apigateway:GET",
+	}
+
+	// This is a tag key unique to this instance of the deployed stack.
+	// Any resource with this unique tag will inherently be scoped to this stack.
+	// This is used to scope the permissions of the lambda to only resources created by this stack.
+	// stackScopedNameKey := tags.GetResourceNameKey(a.stackId)
```
Comment (HomelessDinosaur) [inline cloud/aws/deploy/batch.go:242]:
```suggestion
```

File: cloud/aws/deploy/batch.go [hunk 5]
```diff
@@ -15,13 +15,348 @@
...
+			{
+				"Action":   listActions,
+				"Effect":   "Allow",
+				"Resource": "*",
+				// "Condition": map[string]map[string]string{
+				// 	// Only apply this to resources who have a resource name key that matches this stack
+				// 	"Null": {
+				// 		fmt.Sprintf("aws:ResourceTag/%s", stackScopedNameKey): "false",
+				// 	},
+				// },
```
Comment (HomelessDinosaur) [inline cloud/aws/deploy/batch.go:258]:
```suggestion
```

File: cloud/aws/deploy/batch.go [hunk 6]
```diff
@@ -15,13 +15,348 @@
...
+		return err
+	}
+
+	_, err = iam.NewRolePolicy(ctx, name+"ListAccess", &iam.RolePolicyArgs{
+		Role:   p.BatchRoles[name].ID(),
+		Policy: pulumi.String(tmpJSON),
+	}, opts...)
+
+	// Create a new Iam Role for the job
+
```
Comment (HomelessDinosaur) [inline cloud/aws/deploy/batch.go:272]:
```suggestion
```

File: cloud/aws/deploy/batch.go [hunk 7]
```diff
@@ -15,13 +15,348 @@
...
+						Name:  "NITRIC_STACK_ID",
+						Value: p.StackId,
+					},
+					{
+						Name:  "AWS_REGION",
+						Value: p.Region,
+					},
+				},
+				JobRoleArn: jobRoleArn,
+				// ExecutionRoleArn: batchRoleArn,
```
Comment (HomelessDinosaur) [inline cloud/aws/deploy/batch.go:326]:
```suggestion
```

File: cloud/aws/deploy/batch.go [hunk 8]
```diff
@@ -15,13 +15,348 @@
...
+			}
+
+			return string(containerPropertiesJson), nil
+		}).(pulumi.StringOutput)
+
+		p.JobDefinitions[jobName], err = batch.NewJobDefinition(ctx, jobName, &batch.JobDefinitionArgs{
+			Name:                pulumi.Sprintf("%s-job-%s", p.StackId, job.Name),
+			ContainerProperties: containerProperties,
+
+			// TODO: Set tags for job definition discovery
```
Comment (HomelessDinosaur) [inline cloud/aws/deploy/batch.go:355]: Has this been done?

File: cloud/aws/runtime/batch/batch.go [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+	awsRegion := env.AWS_REGION.String()
+
+	// Create a new AWS session
+	cfg, sessionError := config.LoadDefaultConfig(context.TODO(), config.WithRegion(awsRegion), config.WithClientLogMode(aws.LogRetries|aws.LogResponse|aws.LogRequest))
+	if sessionError != nil {
+		return nil, fmt.Errorf("error creating new AWS session %w", sessionError)
+	}
+
+	return &AwsBatchService{
+		// TODO: Configure client
```
Comment (HomelessDinosaur) [inline cloud/aws/runtime/batch/batch.go:98]: How do you want the client configured?

File: cloud/aws/common/config.go [hunk 10]
```diff
@@ -23,19 +23,28 @@ import (
...
+	MaxCpus       int      `mapstructure:"max-cpus"`
+	InstanceTypes []string `mapstructure:"instance-types"`
 }
 
 type AwsConfig struct {
 	ScheduleTimezone                      string `mapstructure:"schedule-timezone,omitempty"`
 	Import                                AwsImports
 	Refresh                               bool
 	Apis                                  map[string]*AwsApiConfig
+	BatchComputeEnvConfig                 *BatchComputeEnvConfig `mapstructure:"batch-compute-env,omitempty"`
```
Comment (HomelessDinosaur) [inline cloud/aws/common/config.go:47]: The GCP config uses the key:`batch-compute`. Should we change this for consistency?

```suggestion
	BatchComputeEnvConfig                 *BatchComputeEnvConfig `mapstructure:"batch-compute,omitempty"`
```

File: cloud/aws/deploy/batch.go (hunk 5 above)
Comment (tjholm) [reply cloud/aws/deploy/batch.go:258]: leaving this here as a reminder to resolve this if we can, its mostly a documented example of something that didn't work but illustrates the issue that needs to be resolved

File: cloud/aws/deploy/batch.go (hunk 1 above)
Comment (tjholm) [reply cloud/aws/deploy/batch.go:74]: Possibly, I don't have a good solution for it, but it should be fine for preview.

File: cloud/aws/common/config.go (hunk 10 above)
Comment (tjholm) [reply cloud/aws/common/config.go:47]: for GCP its a bit different, the `ComputeEnvironment` is a proper resource in AWS. Happy to look at updating the name in GCP instead. But wanted the key to be a reflection of the resource that was being created in the users AWS account.

**Review by HomelessDinosaur — APPROVED** [review]

**Review by jyecusch — APPROVED** [review]

**Pull request merged**